tracer/src/tracer.rs

```diff
@@ -0,0 +1,170 @@
+// Copyright © 2022 Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use once_cell::unsync::OnceCell;
+use serde::Serialize;
+use std::collections::HashMap;
+use std::fs::File;
+use std::io::Write;
+use std::sync::atomic::{AtomicU64, Ordering};
+use std::sync::{Arc, Mutex};
+use std::time::{Duration, Instant};
+
+#[derive(Debug)]
+struct Tracer {
+    events: Arc<Mutex<HashMap<String, Vec<TraceEvent>>>>,
+    thread_depths: HashMap<String, Arc<AtomicU64>>,
+    start: Instant,
+}
+
+impl Tracer {
+    fn new() -> Self {
+        Self {
+            events: Arc::new(Mutex::new(HashMap::default())),
+            start: Instant::now(),
+            thread_depths: HashMap::default(),
+        }
+    }
+
+    fn end(&self) {
+        let end = Instant::now();
+        let path = format!("cloud-hypervisor-{}.trace", unsafe { libc::getpid() });
+        let mut file = File::create(&path).unwrap();
+
+        #[derive(Serialize)]
+        struct TraceReport {
+            duration: Duration,
+            events: Arc<Mutex<HashMap<String, Vec<TraceEvent>>>>,
+        }
+
+        let trace_report = TraceReport {
+            duration: end.duration_since(self.start),
+            events: self.events.clone(),
+        };
+
+        serde_json::to_writer_pretty(&file, &trace_report).unwrap();
+
+        file.flush().unwrap();
+
+        warn!("Trace output: {}", path);
+    }
+
+    fn add_event(&mut self, event: TraceEvent) {
+        let thread_name = std::thread::current().name().unwrap_or("").to_string();
+        let mut events = self.events.lock().unwrap();
+        if let Some(thread_events) = events.get_mut(&thread_name) {
+            thread_events.push(event);
+        } else {
+            events.insert(thread_name.to_string(), vec![event]);
+        }
+    }
+
+    fn increase_thread_depth(&mut self) {
+        let thread_name = std::thread::current().name().unwrap_or("").to_string();
+        if let Some(depth) = self.thread_depths.get_mut(&thread_name) {
+            depth.fetch_add(1, Ordering::SeqCst);
+        } else {
+            self.thread_depths
+                .insert(thread_name, Arc::new(AtomicU64::new(0)));
+        }
+    }
+
+    fn decrease_thread_depth(&mut self) {
+        let thread_name = std::thread::current().name().unwrap_or("").to_string();
+        if let Some(depth) = self.thread_depths.get_mut(&thread_name) {
+            depth.fetch_sub(1, Ordering::SeqCst);
+        } else {
+            panic!("Unmatched decreas for thread: {}", thread_name);
+        }
+    }
+
+    fn thread_depth(&self) -> u64 {
+        let thread_name = std::thread::current().name().unwrap_or("").to_string();
+        self.thread_depths
+            .get(&thread_name)
+            .map(|v| v.load(Ordering::SeqCst))
+            .unwrap_or_default()
+    }
+}
+
+static mut TRACER: OnceCell<Tracer> = OnceCell::new();
```

Comment:
Just curious why you're using `once_cell::unsync::OnceCell` rather than `once_cell::sync::Lazy` given it's already used through our codebase.

Reply:
`Lazy` is great if you want to initialise it with a value at some point in time during the program creation. I however want to create the `Tracer` only at start time.